File: WWIDesigner/legacy/ImpedanceSpectrum.cc
// FluteCalc
// Copyright Dan Gordon, 2006

#include "ImpedanceSpectrum.hh"

void CalcImpedance(ImpedanceSpectrum & impSpect, Flute & flute, 
		double freq_start, double freq_end, int nfreq)
{
	complex prev_z = ZERO;
	double abs_prev_prev_z = 0;
	double prev_freq = 0;
	double freq_step = (freq_end - freq_start) / (nfreq - 1);
	for (int i = 0; i < nfreq; ++i)
	{
		double freq = freq_start + i * freq_step;
		complex z_ac = flute.CalcZ(freq);
		double abs_z_ac = abs(z_ac);

		impSpect.SetDataPoint(freq, z_ac);

		double abs_prev_z = abs(prev_z);

		if ((i >= 2) && 
				(abs_prev_z < abs_z_ac) && 
				(abs_prev_z < abs_prev_prev_z))
		{
			// We have found an impedance minimum.
			impSpect.Minima().push_back(prev_freq);
		}

		if ((i >= 2) && 
				(abs_prev_z > abs_z_ac) && 
				(abs_prev_z > abs_prev_prev_z))
		{
			// We have found an impedance maximum.
			impSpect.Maxima().push_back(prev_freq);
		}

		abs_prev_prev_z = abs_prev_z;
		prev_z = z_ac;
		prev_freq = freq;
	}
}

File: WWIDesigner/src/main/com/wwidesigner/math/ImpedanceSpectrum.java
/**
 * 
 */
package com.wwidesigner.math;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.TreeMap;

import org.apache.commons.math3.complex.Complex;

import com.wwidesigner.geometry.InstrumentInterface;

/**
 * Representation of a complex spectrum, along with information about its
 * extreme points.
 */
public class ImpedanceSpectrum
{

    private Map<Double, Complex> mSpectrum = new TreeMap<Double, Complex>();
    private List<Double> mMinima = new ArrayList<Double>();
    private List<Double> mMaxima = new ArrayList<Double>();

    /**
     * Add or replace a point in the spectrum.
     */
    public void setDataPoint( double frequency, Complex impedance )
    {
        mSpectrum.put( frequency, impedance );
    }

    public void calcImpedance( InstrumentInterface flute, double freqStart, double freqEnd, int nfreq )
    {
        Complex prevZ = Complex.ZERO;
        double absPrevPrevZ = 0;
        double prevFreq = 0;
        double freqStep = ( freqEnd - freqStart ) / ( nfreq - 1 );
        for ( int i = 0; i < nfreq; ++i )
        {
            double freq = freqStart + i * freqStep;
            Complex zAc = flute.calcZ( freq );
            double absZAc = zAc.abs();

            setDataPoint( freq, zAc );

            double absPrevZ = prevZ.abs();

            if ( ( i >= 2 ) && ( absPrevZ < absZAc ) && ( absPrevZ < absPrevPrevZ ) )
            {
                // We have found an impedance minimum.
                getMinima().add( prevFreq );
            }

            if ( ( i >= 2 ) && ( absPrevZ > absZAc ) && ( absPrevZ > absPrevPrevZ ) )
            {
                // We have found an impedance maximum.
                getMaxima().add( prevFreq );
            }

            absPrevPrevZ = absPrevZ;
            prevZ = zAc;
            prevFreq = freq;
        }
    }

    public List<Double> getMaxima()
    {
        return mMaxima;
    }

    public void setMaxima( List<Double> maxima )
    {
        mMaxima = maxima;
    }

    public List<Double> getMinima()
    {
        return mMinima;
    }

    public void setMinima( List<Double> minima )
    {
        mMinima = minima;
    }

    public Map<Double, Complex> getSpectrum()
    {
        return mSpectrum;
    }

    public void setSpectrum( Map<Double, Complex> spectrum )
    {
        mSpectrum = spectrum;
    }
}
